src/helpers/buffers/unordered_receiver.rs

```diff
@@ -0,0 +1,502 @@
+    /// be entered into normal `wakers` correctly.  Those that are too far
+    /// ahead (i.e., `d = i - (next + c)` is large) they will be woken at most
+    /// `1 + 2d/c` times extra.
+    ///
+    /// Assuming that the awoken items are polled in a timely fashion, this
+    /// ensures that any overflow will be registered in `wakers` (or read)
+    /// before the data needs to be read.  Notifying every `c` items rather
+    /// than `c/2` could mean that a task is not able to poll and enter
+    /// `wakers` in time to be read out.
+    ///
+    /// Note: in protocols we try to send before receiving, so we can rely on
+    /// that easing load on this mechanism.  There might also need to be some
+    /// end-to-end back pressure for tasks that do not involve sending at all.
+    overflow_wakers: Vec<Waker>,
+    /// Left over data from a previous `recv()` call.  The underlying stream
+    /// can provide chunks of data larger than a single message.  We save the
+    /// spare data here.
+    spare: Spare,
+    /// The absolute index of the next value that will be received.
+    next: usize,
+    _marker: PhantomData<C>,
+}
+
+impl<S, C> OperatingState<S, C>
+where
+    S: Stream<Item = C>,
+    C: AsRef<[u8]>,
+{
+    /// Determine whether `i` is the next record that we expect to receive.
+    fn is_next(&self, i: usize) -> bool {
+        i == self.next
+    }
+
+    /// Track a waker from a future that was invoked before data was ready.
+    ///
+    /// Note that this only tracks the last waker for each index, which appears
+    /// to violate the contract for `Future` implementations.  That shouldn't a
+    /// problem in practice because we expect [`recv`] to be called only once,
+    /// which produces just one `Future`.  Informing the last context to [`poll`]
+    /// that `Future` should at least ensure that things progress, even if we
+    /// don't guarantee that all contexts are awoken.
+    ///
+    /// # Panics
+    ///
+    /// If `i` is for an message that has already been read.
+    ///
+    /// [`recv`]: UnorderedReceiver::recv
+    /// [`poll`]: Future::poll
+    fn add_waker(&mut self, i: usize, waker: Waker) {
+        // We don't save a waker at `self.next`, so `>` and not `>=`.
+        if i > self.next + self.wakers.len() {
+            self.overflow_wakers.push(waker);
+        } else {
+            assert!(
+                i > self.next,
+                "Awaiting a read that has already been fulfilled"
+            );
+            let index = i % self.wakers.len();
+            self.wakers[index] = Some(waker);
+        }
+    }
+
+    /// Wake the waker from the next future, if the next receiver has been polled.
+    fn wake_next(&mut self) {
+        self.next += 1;
+        let index = self.next % self.wakers.len();
+        if let Some(w) = self.wakers[index].take() {
+            w.wake();
+        }
+        // See the documentation for `overflow_wakers`.
+        if self.next % (self.wakers.len() / 2) == 0 {
```

Comment:
Well, *this* design depends on this being done unconditionally, but I think I'm going to try your suggested design instead.

Reply:
ah I think I got it now. You wake the overflow wakers not just because there might be a chance that their future is ready, but also because you want to move some of them to the `wakers` vector. In this case you have to wake them up on schedule